Full editorial pass (#111)
ietf-wg-masque/draft-ietf-masque-quic-proxy: draft-ietf-masque-quic-proxy.md

and `0000CCCC` as the proxy CID. The client has connected to a single target
through the proxy, and the client-to-target QUIC connection has `CCCC1111`
as the client CID and `1111CCCC` as the target CID. In order to use
forwarded mode, the proxy assigns VCIDs to use on the client-to-proxy link
to represent the client-to-target connection. In this case, the VCIDs could
be `AAAABBBB` for the client's VCID and `BBBBAAAA` for the proxy's VCID that
it uses to forward to the target.

~~~ aasvg
   +--------+              +-----------+              +--------+
   |        |              |           |              |        |
   | Client +--------------+   Proxy   +--------------+ Target |
   |        |              |           |              |        |
   +--------+              +-----------+              +--------+

       CCCC0000       0000CCCC
       <-- Client-to-proxy -->
              connection

       CCCC1111                                     1111CCCC
       <----------- Client-to-target connection ----------->

  AAAABBBB (CCCC1111)   BBBBAAAA (1111CCCC)
       <-- Client-to-proxy -->
                VCIDs
~~~

> {: #vcid-diagram title="Diagram of VCIDs in forwarded mode"}

Reviewer: It seems like we're jumping between a few different conventions for CIDs and I'm wondering if there's an opportunity for more consistency.

In the "Example of sharing a proxy-to-target 4-tuple" diagram prior to this, I see the following for the client<->target connection:
- 4 of the same letter (the client's letter) followed by the same 4 numbers for client CID
- Those same 4 letters followed by those 4 numbers all incremented by one

In this diagram, we have to show CIDs for three different connections which requires 6 different CIDs. The convention switches to doing 4 letters followed by 4 numbers vs. those same 4 numbers followed by four letters.  I think having a distinct pattern for each of these that's consistent across diagrams might be useful.

I think the CCCC0000, 0000CCCC CID pair pattern may be the least distracting. Each time there's a new client, go to the next letter in the alphabet. Each time a single client needs another CID pair, increment the number. For this diagram, that would mean instead of AAAABBBB we would have CCCC2222

Author: Updated!

In order for a proxy to correctly route packets using VCIDs from
client-to-target and target-to-client, the proxy MUST guarantee that
the mappings between VCIDs, CIDs, and 4-tuples are unique. Specifically,
in order to route packets sent by the client, the proxy needs to be able
to observe the VCID and the client-to-proxy 4-tuple, and map them
to a specific target CID and proxy-to-target 4-tuple. In order to route

> packets send by a target, the proxy needs to be able to observe the client

Reviewer: ```suggestion
packets sent by a target, the proxy needs to be able to observe the client
```

CID and the proxy-to-target 4-tuple, and map them to a specific VCID
and client-to-proxy 4-tuple. Since proxies choose the VCID values, they
can ensure that the VCIDs are distinguishable.

Servers receiving QUIC packets can employ load balancing
strategies such as those described in {{?QUIC-LB=I-D.ietf-quic-load-balancers}}
that encode routing information in the connection ID. When operating in
forwarded mode, clients send QUIC packets destined for the target directly

> to the proxy. Since these packets are generated using the target CID,
> load balancers would not be able to route packets to the correct proxy if the
> packets were sent with the target CID. The target VCID

Reviewer: ```suggestion
to the proxy. Since these packets are generated using the target CID,
load balancers may not have the necessary information to route packets to the correct proxy. The target VCID
```

within the client-to-proxy connection.

## Negotiating Modes and Connection IDs {#negotiation}

In order to support QUIC-aware proxying, both clients and proxies need
to support capsules {{HTTP-CAPSULES}}, which is indicated by including
the "Capsule-Protocol" header field in requests and responses. If this header
field is not included, none of the functionality in this document can be used.

~~~
  capsule-protocol = ?1
~~~

To support forwarded mode, both clients and proxies need to include
the "Proxy-QUIC-Forwarding" header field, as defined in {{forwarding-header}}.
This indicates support for forwarded mode, and allows negotiation of
packet transforms to apply when forwarding ({{transforms}}), along
with any parameters for specific transforms. Clients or proxies that
don't support forwarded mode will not include this header field.

~~~
  proxy-quic-forwarding = ?1; accept-transform=scramble,identity; \
      scramble-key=:abc...789=:
~~~

After negotiating support with header fields, clients and proxies use
the capsules defined in {{cid-capsules}} to communicate information
about CIDs and VCIDs.

For QUIC-aware proxying without forwarded mode, the steps are as follows:

> 1. The client sends the `REGISTER_CLIENT_CID` capsule as soon as it starts

Reviewer: This numbered list all starts with `1.`

Author: Yes that's how markdown numbered lists work, it's weird

sending packets to the target that define that CID.

1. The proxy sends the `ACK_CLIENT_CID` capsule to acknowledge that CID,
with no associated client VCID; alternatively, the proxy can send the
`CLOSE_CLIENT_CID` if it detects a conflict with another CID.

1. The client sends the `REGISTER_TARGET_CID` capsule as soon as it learns
the target CID on the client-to-target connection.

1. The proxy sends the `ACK_TARGET_CID` capsule to acknowledge that CID,
with no associated target VCID; alternatively, the proxy can send the
`CLOSE_TARGET_CID` if it detects a conflict with another CID.

1. Whenever a client or target stops uses a particular CID, the client
sends a `CLOSE_CLIENT_CID` or `CLOSE_TARGET_CID` capsule. The client
can also initiate new `REGISTER_CLIENT_CID` or `REGISTER_TARGET_CID`
exchanges at any time.

> For QUIC-aware proxying with forwarded mode, the steps are as follows:

Reviewer: The steps are nearly identical between the two modes - should we consolidate?

Author: They're close, but there are differences in almost all steps for VCID vs no VCID

The client MUST add an "accept-transform" parameter whose value is an
`sf-string` containing the supported packet transforms ({{transforms}})
in order of descending preference, separated by commas. If the proxy receives a
"Proxy-QUIC-Forwarding" header without the "accept-transform" parameters, it
MUST ignore the header and respond as if the client had not sent the
"Proxy-QUIC-Forwarding" header.

If the proxy supports QUIC-aware proxying, it will include the
"Proxy-QUIC-Forwarding" header in successful HTTP responses. The value
indicates whether or not the proxy supports forwarding. If the client does
not receive this header in responses, the client SHALL assume that the proxy
does not support this extension.

The proxy MUST include a "transform" parameter whose value is an `sf-string`
indicating the selected transform. If the proxy does not recognize or accept
any of the transforms offered by the client, it MUST omit this parameter and
set the header field value to "?0", or omit the header entirely.

# Connection ID Capsules {#cid-capsules}

Connection ID awareness relies on using capsules {{HTTP-CAPSULES}} to
signal addition and removal of Connection IDs. Clients send capsules

> to let proxies know when new Connection IDs on the client-to-target
> QUIC connection are changed. Proxies send capsules to acknowledge or

Reviewer: ```suggestion
to let proxies know when Connection IDs on the client-to-target
QUIC connection are changing. Proxies send capsules to acknowledge or
```
